src/middlewared/middlewared/plugins/idmap.py

```diff
@@ -23,281 +24,342 @@ class dstype(enum.Enum):
-            if key in data and data[key] is not None:
-                data[key] = data[key].upper()
+        opt_enums = ['ssl', 'linked_service']
+        if data.get('options'):
+            for i in opt_enums:
+                if data['options'].get(i):
+                    data['options'][i] = data['options'][i].upper()
 
         return data
 
     @private
-    async def common_backend_compress(self, data):
-        for key in ['ldap_server', 'schema_mode', 'ssl']:
-            if key in data and data[key] is not None:
-                data[key] = data[key].lower()
-
-        if 'id' in data['domain'] and data['domain']['id']:
-            domain_info = await self.middleware.call(
-                'idmap.domain.query', [('id', '=', data['domain']['id'])]
-            )
-            data['domain'] = domain_info[0]['name']
-        elif 'idmap_domain_name' in data['domain'] and data['domain']['idmap_domain_name']:
-            data['domain'] = data['domain']['idmap_domain_name']
-        else:
-            domain_info = await self.middleware.call(
-                'idmap.domain.query', [('domain', '=', data['domain']['idmap_domain_dns_domain_name'].upper())]
-            )
-            data['domain'] = domain_info[0]['name']
+    async def idmap_compress(self, data):
+        opt_enums = ['ssl', 'linked_service']
+        if data.get('options'):
+            for i in opt_enums:
+                if data['options'].get(i):
+                    data['options'][i] = data['options'][i].lower()
+
+        data['idmap_backend'] = data['idmap_backend'].lower()
 
         return data
 
     @private
-    async def _validate_domain_info(self, domain, verrors):
+    async def get_next_idmap_range(self):
         """
-        Only domains that have been configured as idmap domains
-        are permitted.
+        Increment next high range by 100,000,000 ids. This number has
+        to accomodate the highest available rid value for a domain.
+        Configured idmap ranges _must_ not overlap.
         """
-        configured_domains = await self.middleware.call('idmap.domain.query')
-
-        id = domain.get('id')
-        id_verified = False
+        domains = await self.query()
+        sorted_idmaps = sorted(domains, key=lambda domain: domain['range_high'])
+        low_range = sorted_idmaps[-1]['range_high'] + 1
+        high_range = sorted_idmaps[-1]['range_high'] + 100000000
+        return (low_range, high_range)
 
-        short_name = domain.get('idmap_domain_name')
-        short_name_verified = False
+    @private
+    async def remove_winbind_idmap_tdb(self):
+        sysdataset = (await self.middleware.call('systemdatast.config'))['basename']
+        ts = str(datetime.datetime.now(datetime.timezone.utc).timestamp())[:10]
+        await self.middleware.call('zfs.snapshot.create', {'dataset': f'{sysdataset}/samba4',
+                                                           'name': f'wb_cache-{ts}'})
+        try:
```

Comment:
Any point in verifying the file exists first so we dont eventually get tracebacks? Or it should always be there?

Reply:
I will log traceback only if we get an error that's not FileNotFound. winbindd_idmap.tdb may or may not be present (typically is).